### PR TITLE
Support arbitrary generating terms in pulse Hamiltonians with `stoch_pulse_grad`

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
Thus far, the gradient transform `stoch_pulse_grad` only supports Pauli words as generating terms in the pulse Hamiltonian.

**Description of the Change:**
This PR generalizes the support to arbitrary generating terms, using the existing generalized shift rule framework.
The inserted operations are no longer `PauliRot` gates but `Exp` operators, because for non-Pauli word generators, the decomposition into Pauli rotation gates is not always desired.
Decomposing `Exp` operations with Pauli words in the exponent yields back the `PauliRot` operations used by `stoch_pulse_grad` until now.

**Benefits:**
Support many more pulse Hamiltonians, in particular e.g. neutral atom Hamiltonians.

**Possible Drawbacks:**
Breaking change: For Pauli word-generated terms, the inserted operations are now `Exp` operations instead of `PauliRot` gates.

**Open TODOs**
- [x] Add example of non-Pauli word generator to docstring.
- [x] Add JIT-compatibility test with non-Pauli word generator.

[sc-34236]

### CODE REVIEW

File: pennylane/ops/op_math/symbolicop.py [hunk 1]
```diff
@@ -196,6 +196,16 @@ def _check_and_compute_batch_size(self, scalar):
             )
         return scalar_size
 
+    @property
+    def hash(self):
```
Comment (dwierichs) [inline pennylane/ops/op_math/symbolicop.py:200]: Without this, `Exp(a * Op)` and `Exp(b* Op)` have the same hash.

**Review by Qottmann — COMMENTED** [review]
Very nice upgrade @dwierichs 💪 

Much appreciate the developer comments 🤩 

There is a slight performance regression on the first execution as `ParametrizedHamiltonian` gets jit compiled inside the ODE solver (because the ode `fun` is jit-compiled), most likely due to the extra steps from `qml.exp`, so dont think there is much we can do. Could short-cut the execution by checking whether it is a pauli word, but that might make the code significantly uglier.

Execution time is unchanged 👍

**Comment by dwierichs** [comment]
>Could short-cut the execution by checking whether it is a pauli word, but that might make the code significantly uglier.

I tried this and it seems to be fine to me, actually :innocent: Let me know what you think :)

**Review by Qottmann — APPROVED** [review]
Very neat upgrade with not many changes in code and tests, really nice work @dwierichs 👌 

With the upcoming short-cutting for Pauli-words looks good to me 🙂

**Comment by dwierichs** [comment]
@Qottmann while fixing the bug in #4156 I found a simple way to extract the prefactor of a Pauli word that I didn't see before. I used this to implement the `PauliRot` shortcut for Pauli words in this PR :)
This includes single-term `Hamiltonian`s, `SProd` instances and (tensor products of) simple Pauli operators.

File: pennylane/gradients/pulse_gradient.py [hunk 2]
```diff
@@ -55,55 +56,65 @@ def _assert_has_jax(transform_name):
...
+    # If we are broadcasting, make use of the `return_intermediate` and `complementary` features
+    ops = tuple(
+        [
+            op(op.data, before_t, return_intermediate=bcast, **ode_kwargs),
+            insert_op,
+            op(op.data, after_t, return_intermediate=bcast, complementary=bcast, **ode_kwargs),
+        ]
+        for insert_op in insert_ops
+    )
+    return ops, qml.math.array(coeffs)
```
Comment (rmoyard) [inline pennylane/gradients/pulse_gradient.py:114]: Here I would use jax directly for the coeffs, we had some trouble in the past with qml.math.array causing troubles with backprop.

File: pennylane/gradients/pulse_gradient.py [hunk 3]
```diff
@@ -152,23 +165,34 @@ def _parshift_and_integrate(
...
+        def _psr_and_contract(res_list, cjacs, int_prefactor):
+            res = qml.math.stack(res_list)
+            # Reshape the results such that the first axis corresponds to the splitting times
+            # and the second axis corresponds to different shifts. All other axes are untouched
+            shape = qml.math.shape(res)
+            new_shape = (shape[0] // num_shifts, num_shifts) + shape[1:]
+            res = qml.math.reshape(res, new_shape)
+            # Contract the results with the parameter-shift rule coefficients
+            parshift = qml.math.tensordot(psr_coeffs, res, axes=[[0], [1]])
+            return qml.math.tensordot(parshift, cjacs, axes=[[0], [0]]) * int_prefactor
```
Comment (rmoyard) [inline pennylane/gradients/pulse_gradient.py:188]: Interesting that the batching case is simpler? What is the reason we don't need to reshape in the broadcasting case?

File: pennylane/gradients/pulse_gradient.py [hunk 4]
```diff
@@ -300,7 +324,8 @@ def sin(p, t):
             return jax.numpy.sin(p * t)
 
         ZZ = qml.PauliZ(0) @ qml.PauliZ(1)
-        H = 0.5 * qml.PauliX(0) + qml.pulse.constant * ZZ + sin * qml.PauliX(1)
+        Y_plus_X = qml.dot([1/5, 3/5], [qml.PauliY(0), qml.PauliX(1)])
+        H = 0.5 * qml.PauliX(0) + qml.pulse.constant * ZZ + sin * Y_plus_X
```
Comment (rmoyard) [inline pennylane/gradients/pulse_gradient.py:328]: :100:

File: tests/gradients/core/test_pulse_gradient.py [hunk 5]
```diff
@@ -87,15 +163,15 @@ def test_output_properties(self, test_case):
             _ops[2].t = op.t
             assert qml.equal(_ops[2], op)
 
-            # Check that the inserted Pauli rotation is correct
-            assert isinstance(_ops[1], qml.PauliRot)
-            assert qml.math.allclose(_ops[1].data, sign * np.pi / 2)
-            assert _ops[1].hyperparameters["pauli_word"] == word
+            # Check that the inserted exponential is correct
+            print(exp_shift)
+            print(_ops[1].data)
```
Comment (rmoyard) [inline tests/gradients/core/test_pulse_gradient.py:168]:
```suggestion
```

**Review by rmoyard — COMMENTED** [review]
Couple of questions and suggestion, otherwise it looks good :100:

File: tests/gradients/core/test_pulse_gradient.py [hunk 6]
```diff
@@ -43,23 +43,75 @@ class TestSplitEvolOps:
...
         (
             ham_two_q_pwc,
             [np.linspace(0, 1, 13)],
             (0.2, 0.6, 0.9, 1.8),
             qml.PauliY(0) @ qml.PauliX(1),
+            "YX",
+        ),
+        (ham_single_q_const, [0.3], 2.3, qml.Hamiltonian([0.2], [qml.PauliZ(0)]), "Z"),
+        (ham_single_q_const, [0.3], 2.3, 1.2 * qml.PauliZ(0), "Z"),
+        (ham_single_q_const, [0.3], 2.3, qml.dot([1.9], [qml.PauliZ(0)]), "Z"),
```
Comment (Qottmann) [inline tests/gradients/core/test_pulse_gradient.py:58]: Would it make sense to also add explicit cases with `qml.s_prod` and `qml.sum`, in case the default behavior for ops with `*` and `+` changes?

File: pennylane/gradients/pulse_gradient.py [hunk 7]
```diff
@@ -90,19 +90,26 @@ def _split_evol_ops(op, ob, tau):
         before_t = jax.numpy.array([t0, tau])
         after_t = jax.numpy.array([tau, t1])
 
-    eigvals = qml.eigvals(ob)
-    coeffs, shifts = zip(*generate_shift_rule(eigvals_to_frequencies(tuple(eigvals))))
+    if qml.pauli.is_pauli_word(ob):
+        prefactor = next(iter(qml.pauli.pauli_sentence(ob).values()))
```
Comment (Qottmann) [inline pennylane/gradients/pulse_gradient.py:94]: lol nice!

Comment (Qottmann) [reply pennylane/gradients/pulse_gradient.py:94]: Could consider adding functionality to obtain this directly somehow in the pauli module 😅

**Review by Qottmann — APPROVED** [review]
Thanks for fixing the short-cut for pauli words 👌 Jit times look much better compared to current master for those cases. Execution time is the same 👍

File: pennylane/gradients/pulse_gradient.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/gradients/pulse_gradient.py:94]: Yeah... I'll ask around whether this is needed elsewhere.

File: pennylane/gradients/pulse_gradient.py (hunk 2 above)
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:114]:
```suggestion
    return ops, jnp.array(coeffs)
```

File: pennylane/gradients/pulse_gradient.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/gradients/pulse_gradient.py:188]: It's because of the ordering of the tapes: With broadcasting, we get all the results that belong to different splitting times $\tau$ in consecutive array entries. Without broadcasting, we get all results for a specific $\tau$ together, then for the next $\tau$ etc...

Comment (dwierichs) [reply pennylane/gradients/pulse_gradient.py:188]: In particular, with broadcasting we get results in 2d `[batch_first_shift, batch_second_shift,...]` where each `batch.._shift` is a 1d array. Without broadcasting we get a flat list `[first_shift_first_tau, second_shift_first_tau..,first_shift_second_tau...]`

File: tests/gradients/core/test_pulse_gradient.py (hunk 6 above)
Comment (dwierichs) [reply tests/gradients/core/test_pulse_gradient.py:58]: Sure, can do :)

File: pennylane/gradients/pulse_gradient.py (hunk 3 above)
Comment (rmoyard) [reply pennylane/gradients/pulse_gradient.py:187]: Thanks :100:

**Review by rmoyard — APPROVED** [review]
Thanks a lot! :+1:
